File: 789/go/func1.go
package main

import "fmt"

func LeftBorder(nums []int, target int) int {
	left, right := 0, len(nums)-1
	for left < right {
		mid := left + ((right - left) >> 1)
		if nums[mid] >= target {
			right = mid
		} else {
			left = mid + 1
		}
	}
	if nums[left] == target {
		return left
	} else {
		return -1
	}
}

func RightBorder(nums []int, target int) int {
	left, right := 0, len(nums)-1
	for left < right {
		mid := left + ((right - left) >> 1) + 1
		if nums[mid] <= target {
			left = mid
		} else {
			right = mid - 1
		}
	}
	if nums[right] == target {
		return right
	} else {
		return -1
	}
}

func main() {
	n, q, k := 0, 0, 0
	fmt.Scanf("%d%d", &n, &q)
	nums := make([]int, 0)
	for i := 0; i < n; i++ {
		x := 0
		fmt.Scan(&x)
		nums = append(nums, x)
	}
	for q != 0 {
		q--
		fmt.Scanf("%d", &k)
		fmt.Printf("%d %d\n", LeftBorder(nums, k), RightBorder(nums, k))
	}
}

File: 789/cpp/func1.cpp
// func 1
// Use 2 functions to find left & right border
#include<iostream>
#include<cstdio>
#include<vector>
using namespace std;

const int N = 1e5 + 10;
int n = 0;
int q = 0;
int k = 0;

int LeftBorderBinarySearch(vector<int>& nums, int target) {
    int left = 0, right = nums.size() - 1;
    while ( left < right ) {
        int mid = left + ((right - left) >> 1);
        if ( nums[mid] >= target ) {
            right = mid;
        } else {
            left = mid + 1;
        }
    }
    return nums[left] == target ? left : -1;
}

int RightBorderBinarySearch(vector<int>& nums, int target) {
    int left = 0, right = nums.size() - 1;
    while ( left < right ) {
        int mid = left + ( (right - left) >> 1 ) + 1;
        if ( nums[mid] <= target ) {
            left = mid;
        } else {
            right = mid - 1;
        }
    }
    return nums[right] == target ? right : -1;
}

int main () {
    vector<int> nums;
    cin >> n >> q;
    for ( int i = 0 ; i < n ; i++ ) {
        int x;
        cin >> x;
        nums.push_back(x);
    }
    while ( q-- ) {
        cin >> k;
        cout << LeftBorderBinarySearch(nums, k) << " " << RightBorderBinarySearch(nums, k) << endl;
    }

    return 0;
}